service/security/shiro/src/main/java/org/eclipse/kapua/service/authentication/shiro/realm/UserPassAuthenticatingRealm.java

```diff
@@ -211,6 +213,20 @@ protected void assertCredentialsMatch(AuthenticationToken authcToken, Authentica
             throws AuthenticationException {
         LoginAuthenticationInfo kapuaInfo = (LoginAuthenticationInfo) info;
         CredentialService credentialService = LOCATOR.getService(CredentialService.class);
+        MfaOptionService mfaOptionService = LOCATOR.getService(MfaOptionService.class);
+        KapuaId userId = kapuaInfo.getUser().getId();
+        KapuaId scopeId = kapuaInfo.getUser().getScopeId();
+        boolean hasMfa = false;
+        boolean userAndPasswordMatch = ((UserPassCredentialsMatcher) getCredentialsMatcher()).doUsernameAndPasswordMatch(authcToken, info);
+        try {
+            if (KapuaSecurityUtils.doPrivileged(() -> mfaOptionService.findByUserId(userId, scopeId) != null)) {
+                hasMfa = true;
+            }
+        } catch (KapuaException e) {
+            e.printStackTrace();
```

Comment:
`e.printStackTrace()` is evil.
Please use SLF4J `Logger` and throw the exception if needed.